@here/harp-materials/lib/SolidLineMaterial.ts

```diff
@@ -568,40 +620,39 @@ export class SolidLineMaterial extends THREE.RawShaderMaterial
 
     /**
      * Size of the gaps between dashed segments.
+     *
+     * @note You may also need to set [[dashSize]].
+     * @see dashSize.
      */
     get gapSize(): number {
         return this.uniforms.gapSize.value as number;
     }
     set gapSize(value: number) {
         this.uniforms.gapSize.value = value;
-        this.defines.DASHED_LINE = this.gapSize > 0.0 ? 1 : 0;
+        setShaderMaterialDefine(this, "DASHED_LINE", value > 0.0 ? 1 : 0);
     }
 
     /**
      * Caps mode.
      */
     get caps(): LineCaps {
         let result: LineCaps = "Round";
-        if (this.defines.CAPS_SQUARE === 1) {
-            result = "Square";
-        } else if (this.defines.CAPS_NONE === 1) {
+        const capsMode = getShaderMaterialDefine(this, "CAPS_MODE");
```

Comment:
MINOR: If we introduce new caps modes we would need to keep this in sync with `LineCapsDefinitions`. Maybe introduce a second definition that does the reverse mapping: `string` -> `LineCapsModes` that is automatically generated from `LineCapsDefinitions`.

Reply:
Sorry @MPursche I don't exactly understand your suggestion, `LineCapsDefinitions` actually does the `string` -> `enum (int)` mapping, using `LineCaps` type (string literals) as key and `LineCapsModes` (enum) as value.
Would you like me to perform this mapping automatically so whenever something is added to `LineCaps` then `LineCapsDefinitions` grow? It is possible, but I am not sure if makes any sense because support in shader need to be added manually - both with new shader const (define), so I doubt if it helps anyway.
On other side if you mean reverse mapping LineCapsModes -> LineCaps (string), I guess it should never be exposed because it is implementation detail of the shaders. Even making `LineCapsModes` exported enum is little bit mistake (should be private) but I needed this in second file where the shader is defined (**LinesChunks.ts**). To fix this I would need to move `LineCapsDefinitions` to **LinesChunks.ts** file, which is maybe even worse....

Reply:
What I meant is, that in `set caps()` the object `LineCapsDefinitions` is used to determine which `LineCaps` string value corresponds to which `LineCapesModes` enum value. For `get caps()` the reverse(`LineCapsModes` to `LineCaps`)is implemented manually with if/else but could also use an object.